### ISSUE TITLE
shopp_captured_order_event with cancelled orders

### ISSUE THREAD

**Issue by dxdc** [ingenesis/shopp#3232]
I've been using the shopp_captured_order_event hook to run a custom function. It works great.

However, in certain cases in the payment logs when I see:

- Order invoiced
- Order canceled    <---- this cancellation seems to mess things up... it works great when this doesn't happen
- Order invoiced
- Payment auth & capture
- Payment authorized
- Payment received

The $Event parameter passed to shopp_captured_order_event doesn't seem to be working for some reason. Or, at least, it's not allowing me to view the order contents.

Any ideas about the cause of this?

**Comment by jonathandavis** [comment]
The Order Canceled happens when the customer submitted the order for processing and at some point they change their order before it goes for payment processing. This usually happens when they land on the Confirm Order page. They may jump off to add something else to their order and it changes the order they just submitted, which changes the order total amount, which needs to change the order balance. Order balance (an amount owed) is set by the `invoiced` order event. So to handle the discrepancy, a `void` (Order canceled) is issued to wipe out the previous balance, and a new `invoiced` event is dispatched to setup the new balance (the correct order total amount owed to the merchant).

It looks like from the event stack this is "working" properly. There may be an issue with resolving the order event chain for access purposes. That would be a bug if we can confirm/reproduce the issue.

**Comment by dxdc** [comment]
Thanks Jonathan... I'm OK with the stack/processing scheme, makes sense, but it seems like the access part isn't working then. Do you know why that would be?

**Comment by jonathandavis** [comment]
No, as I said, we need to reproduce that scenario to confirm the access issue.

**Comment by jonathandavis** [comment]
I couldn't reproduce this with the available details. Is it downloads that you can't view? I was able to view the order just fine, though it does incorrectly show the order as authorized, not paid.

**Comment by dxdc** [comment]
Hi Jonathan, thanks for checking this out! Here are some details: Shopp 1.3.8 running Paypal and Stripe plugins.

Here's the log from the order in question (customer used Stripe):

Yesterday 8:48 am	Order invoiced		$9.95
Yesterday 8:51 am	Order cancelled		
Yesterday 8:51 am	Order invoiced		$9.95
Yesterday 8:51 am	Order cancelled		
Yesterday 8:51 am	Order invoiced		$9.95
Yesterday 8:53 am	Order cancelled		
Yesterday 8:53 am	Order invoiced		$9.95
Yesterday 8:53 am	Payment authorization & capture	


Here's the gist of the function that works normally, but not when there is a series of order invoices/cancels. With just an invoice trigger it works fine.

function mykey($Event)
{
    $Purchase = $Event->order();

    $invoice    = $Purchase->id;
    $user_name  = trim($Purchase->firstname) . ' ' . trim($Purchase->lastname);
    $user_email = strtolower($Purchase->email);
    $product_name_list = array();

    foreach ($Purchase->purchased as $purchased) {

        $Purchased = new Purchased($purchased->id); // loop through each product

        $product_qty  = $Purchased->quantity;
        $product_name = $Purchased->name;

  ....

} 
}

**Comment by dxdc** [comment]
Jonathan, after implementing the JS fix we discussed earlier (double paypal submit), I wasn't able to replicate this anymore. Wonder if that was related to the problem? Will keep an eye on it and report back.

**Comment by jonathandavis** [comment]
More than likely it absolutely was related. The likely scenario that played out here is that the first submit fired and created an orphan order and the second is the one the browser "stayed connected with". By that I mean, the second submission is the one the server's second response is connected to and responds with. This is leading to odd behavior in the way Shopp "cleans up" after the order process. It is leaving the order in the session meaning the continue-editing-order behaviors after the order is submitted are occurring (causing the invoiced-cancelled-invoiced-cancelled-auth chain of events).

Without the second submission, the clean-up process works on the same order that was submitted, so everything is tidied up appropriately and a new order is started after the successful completion of the prior order.

Either way, we may still want to deal with access issues on cancelled-then-completed orders. We'll just have to manually setup that scenario for testing. Either way, this is not a high priority for the Cydonia milestone. We'll more than likely want to revisit this in the future. Dropping this into the Next milestone.

**Comment by dxdc** [comment]
Haven't seen this issue since this fix.
